tools/src/main/scala/scala/scalanative/codegen/MemoryLayout.scala

```diff
@@ -21,23 +21,22 @@ final case class MemoryLayout(
    *  end expresed as number of words. Used by the GC for scanning objects.
    *  Terminated with offset=-1
    */
-  def referenceFieldsOffsets(implicit meta: Metadata): Seq[nir.Val.Long] = {
-    val sizeOfHeader = meta.layouts.ObjectHeader.size
+  private[codegen] def referenceFieldsOffsets(implicit
+      meta: Metadata
+  ): Seq[nir.Val.Int] = {
     import nir.Type._
-    val ptrOffsets =
+    val offsets =
       tys.collect {
         // offset in words without object header
         case MemoryLayout.PositionedType(
               _: RefKind | // normal or alligned reference field
               StructValue((_: RefKind) :: ArrayValue(nir.Type.Byte, _) :: Nil),
               offset
             ) =>
-          nir.Val.Long(
-            (offset - sizeOfHeader) / MemoryLayout.BYTES_IN_LONG
-          )
```

Comment:
This way on calculating reference field indexes could have lead to runtime issues in 32-bit runtime. It would lead to having double indexes (eg. offset=16 & offset=20 would point to the same index) and might omit scanning some of misaligned fields.